hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -524,10 +528,94 @@ def should_generate_more():
         parameter = GenerationParameters(self.random)
         count = 0
 
+        # We use two strategies to choose the maximum test case size.
+        # Ideally we use the precise estimate which is basd on zero extending
+        # the prefix to get what is (probably) the smallest size a valid example
+        # can have. However, if this doesn't work then we have to fall back to
+        # an imprecise estimate, and this can result in a lot of additional
+        # useless calls to the test function. In order to avoid this scenario
+        # we keep track of how many times in a row this happens, and switch
+        # to a strategy of always using the approximate version.
+        consecutive_zero_extend_is_invalid = 0
+
         while should_generate_more():
             prefix = self.generate_novel_prefix()
 
-            data = self.new_conjecture_data(draw_bytes_with(prefix, parameter))
+            # We control growth during initial example generation, for two
+            # reasons:
+            #
+            # * It gives us an opportunity to find small examples early, which
+            #   gives us a fast path for easy to find bugs.
+            # * It avoids low probability events where we might end up
+            #   generating very large examples during health checks, which
+            #   on slower machines can trigger HealthCheck.too_slow.
+            #
+            # The heuristic we use is that we attempt to estimate the smallest
+            # extension of this prefix, and limit the size to no more than
+            # an order of magnitude larger than that. If we fail to estimate
+            # the size accurately, we skip over this prefix and try again.
+            #
+            # We need to tune the example size based on the initial prefix,
+            # because any fixed size might be too small, and any size based
+            # on the strategy in general can fall afoul of strategies that
+            # have very different sizes for different prefixes.
+            small_example_cap = max(10, min(self.settings.max_examples // 10, 50))
+
+            if (
+                self.valid_examples <= small_example_cap
+                and self.call_count <= 5 * small_example_cap
+            ):
+                # If we didn't get a valid example then we can't reliably
+                # use the length to predict the size of the buffer, so we
+                # just take a rough guess.
+                max_length = max(len(prefix) * 10, BUFFER_SIZE // 10)
+
+                if consecutive_zero_extend_is_invalid < 5:
+                    minimal_example = self.cached_test_function(
+                        prefix + hbytes(BUFFER_SIZE)
```

Comment:
Is it OK to submit an input buffer that's longer than `BUFFER_SIZE`?

I doubt anything will go horribly wrong, but it's a bit odd to have a buffer limit that we don't consistently respect, and I'm pretty sure there isn't any automatic truncation between here and the actual test function.

Reply:
Running some quick tests on `master`, it seems that we previously weren't submitting overly-long buffers, *except* in some cases specifically related to the target optimizer.

It's a bit scary to see this property quietly being broken by new code. Again, it's probably not the end of the world, but it probably deserves more attention at some point.

Reply:
Hmm. Good question. It's probably fine but you're right that it's weird. I'll fix that.

Reply:
> It's a bit scary to see this property quietly being broken by new code. Again, it's probably not the end of the world, but it probably deserves more attention at some point.

Shall I add an assertion in to `test_function` that asserts that `len(data.buffer) <= BUFFER_SIZE`?